core/src/main/java/org/infinispan/remoting/transport/MultiTargetRequest.java

```diff
@@ -0,0 +1,164 @@
+package org.infinispan.remoting.transport;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.infinispan.remoting.responses.CacheNotFoundResponse;
+import org.infinispan.remoting.responses.Response;
+import org.infinispan.util.concurrent.TimeoutException;
+
+import net.jcip.annotations.GuardedBy;
+
+/**
+ * Request implementation that waits for responses from multiple target nodes.
+ *
+ * @author Dan Berindei
+ * @since 9.1
+ */
+public class MultiTargetRequest<T> extends AbstractRequest<T> {
+   @GuardedBy("responseCollector")
+   private final Address[] targets;
+   @GuardedBy("responseCollector")
+   private int missingResponses;
+
+   public MultiTargetRequest(ResponseCollector<T> responseCollector, long requestId, RequestRepository repository,
+                             Collection<Address> targets, Address excluded) {
+      super(requestId, responseCollector, repository);
+      this.targets = new Address[targets.size()];
```

Comment:
1. I never know how many objects will be instantiated.
2. Formatting streams-based code on multiple lines is weird, so most of the code I see gives up and puts everything one one line, making it harder to read than the equivalent imperative code.